docs/frontend-development/javascript-guidelines.md

```diff
@@ -0,0 +1,320 @@
+│   ├── frontend.js
+│   └── editor-style-overrides.js
+└── block-editor/
+    ├── blocks/
+    ├── extensions/
+    └── editor.js
+```
+
+Compiled JavaScript is output to the `dist/js/` directory during the build process.
+
+### Module Organization
+
+- Each component should be in its own directory
+- Avoid barrel files (index.js) unless necessary (see: https://vercel.com/blog/how-we-optimized-package-imports-in-next-js)
+- Group related functionality in modules
+- Use named exports for better tree-shaking
+- Keep files small and focused on a single responsibility
+
+Example:
+
+```javascript
+// components/Button/Button.js
+import React from 'react';
+
+export const Button = ({ children, onClick, variant = 'primary' }) => {
+  return (
+    <button
+      className={`button button--${variant}`}
+      onClick={onClick}
+    >
+      {children}
+    </button>
+  );
+};
+
+// Usage in another file
+import { Button } from './components/Button/Button';
+```
+
+## Modern JavaScript Features
+
+### ES6+ Features
+
+Use modern JavaScript features to write cleaner, more maintainable code:
+
+#### Arrow Functions
+
+```javascript
+// Good
+const multiply = (a, b) => a * b;
+
+// Instead of
+function multiply(a, b) {
+  return a * b;
+}
+```
+
+#### Destructuring
+
+```javascript
+// Good
+const { title, content, author } = post;
+
+// Instead of
+const title = post.title;
+const content = post.content;
+const author = post.author;
+```
+
+#### Template Literals
+
+```javascript
+// Good
+const greeting = `Hello, ${name}!`;
+
+// Instead of
+const greeting = 'Hello, ' + name + '!';
+```
+
+#### Spread Operator
+
+```javascript
+// Good
+const newArray = [...oldArray, newItem];
+const newObject = { ...oldObject, newProperty: value };
+
+// Instead of
+const newArray = oldArray.concat([newItem]);
+const newObject = Object.assign({}, oldObject, { newProperty: value });
+```
+
+#### Optional Chaining
+
+```javascript
+// Good
+const userName = user?.profile?.name;
+
+// Instead of
+const userName = user && user.profile && user.profile.name;
+```
+
+#### Nullish Coalescing
+
+```javascript
+// Good
+const count = value ?? 0;
+
+// Instead of
+const count = value !== null && value !== undefined ? value : 0;
+```
+
+## React Best Practices
+
+### Functional Components
+
+Use functional components with hooks instead of class components:
+
+```javascript
+// Good
+import React, { useState, useEffect } from 'react';
+
+const UserProfile = ({ userId }) => {
+  const [user, setUser] = useState(null);
+  const [loading, setLoading] = useState(true);
+
+  useEffect(() => {
+    const fetchUser = async () => {
+      try {
+        const data = await fetchUserData(userId);
+        setUser(data);
+      } catch (error) {
+        console.error('Error fetching user:', error);
+      } finally {
+        setLoading(false);
+      }
+    };
+
+    fetchUser();
+  }, [userId]);
+
+  if (loading) return <Loading />;
+  if (!user) return <Error message="User not found" />;
```

Comment:
Please let us use any other example that doesn't show bad ways of doing data fetching in react 😬

Reply:
TBF, we don't even need to show data loading here. Stripping this bit out